lhotse/dataset/source_separation.py

```diff
@@ -1,41 +1,158 @@
-    cuts for the sources, and the other one the audio cuts for the signal mix. When queried for data samples,
-    it returns a dict of {'sources': tensor, 'mixture': tensor}.
+    An abstract base class, implementing PyTorch Dataset for the source separation task.
+    It's created from two CutSets - one provides the audio cuts for the sources, and the other one the audio cuts for
+    the signal mix. When queried for data samples, it returns a dict of: {
+        'sources': (N x T x F) tensor,
+        'mixture': (T x F) tensor,
+        'real_mask': (N x T x F) tensor,
+        'binary_mask': (T x F) tensor
+    }.
+
+    This Dataset performs on-the-fly feature-domain mixing of the sources. It expects the mixtures_set to contain
+    MixedCuts, so that it knows which Cuts should be mixed together.
     """
 
-    def __init__(self, sources_set: CutSet, mixtures_set: CutSet, root_dir: Optional[Pathlike] = None):
+    def __init__(
+            self,
+            sources_set: CutSet,
+            mixtures_set: CutSet,
+            root_dir: Optional[Pathlike] = None
+    ):
         super().__init__()
         self.sources_set = sources_set
-        self.mixtures_set = mixtures_set
-        self.cut_ids = list(self.mixtures_set.cuts.keys())
+        self.mixtures_set = mixtures_set.with_source_cuts_from(self.sources_set)
         self.root_dir = Path(root_dir) if root_dir else None
-        assert set(self.cut_ids) == set(self.mixtures_set.cuts.keys())
+
+        self.cut_ids = list(self.mixtures_set.cuts.keys())
+        self.validate()
+
+    def _obtain_mixture(self, cut_id: str) -> Tuple[AnyCut, List[Cut]]:
+        raise NotImplementedError("You are using SpeechSeparationDataset, which is an abstract base class; instead, "
+                                  "use one of its derived classes that specify whether the mix is pre-computed or "
+                                  "done dynamically (on-the-fly).")
+
+    def validate(self):
+        for cut in self.mixtures_set.mixed_cuts.values():
+            _, source_cuts = self._obtain_mixture(cut.id)
+            assert len(source_cuts) == 2, f"Source separation with more than two sources is currently not supported. " \
+                                          f"Cut with ID '{cut.id}' has {len(source_cuts)} sources."
 
     def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
         cut_id = self.cut_ids[idx]
-        mixture_cut = self.mixtures_set.cuts[cut_id].with_cut_set(self.sources_set)
-        source_cuts = [self.sources_set.cuts[id_] for id_ in [mixture_cut.left_cut_id, mixture_cut.right_cut_id]]
+        mixture_cut, source_cuts = self._obtain_mixture(cut_id=cut_id)
+
         mixture = torch.from_numpy(mixture_cut.load_features(root_dir=self.root_dir))
         sources = torch.stack(
             [torch.from_numpy(source_cut.load_features(root_dir=self.root_dir)) for source_cut in source_cuts],
             dim=0
         )
+
```

Comment:
I'm not completely sure how it works here when there is noise. Would there be a real_mask for the noise as well? And a source idx in the binary mask for it as well?

Reply:
Hmm, good point. I'm returning the masks as if the noise was not there, which is probably not the best supervision. Would it make sense if I made `real_mask` first dimension equal to the number of sources (+ noise), and changed the `binary_mask` to `source_mask` (not sure about the name)` which uses int values to indicate which bin belongs to which source (e.g. 2 sources - {0, 1} values, 3 sources - {0, 1, 2} values, etc.)? This would also enable this dataset to use source separation for multiple sources.

Reply:
Huh, that's funny, but it actually works like that right now, depending on whether you pass the noise cuts into `sources_set` or `nonsources_set`. When you pass it to `sources_set` (e.g. by doing `source_cut_set + noise_cut_set`), the noise will get treated as a regular source, and it will be involved in the computation of real_mask and "binary"_mask. If you pass it to `nonsources_set` (which was how I intended it for usage) then it will not be returned as one of the sources and won't participate in mask computation - basically, in that case, it only serves to load its features and add them to the mixture.

Which usage would you say makes more sense? BTW I thought the code won't work for more than 2 sources, but it seems like it will, so I'll remove the exception that I'm raising in such cases.

Reply:
Also, as a sanity check, what is a reasonable range of values for `real_mask`? It seems the values I'm getting are quite low (e.g. [-0.002, 0.007]), I'm thinking it's the log that's messing up the mask. We might need to exp these energies first for source separation.

Reply:
I think normally these kinds of masks would be in the range zero to one.
I'd try to make it as flexible as possible and not bake it into the system deeply.  Different kinds of source separation task may need different types of masks, and we should be leaving this open I think.  When someone builds a speech separation setup from this, they can deal with this.

Reply:
Sounds good to me

Reply:
Sounds good to me

Reply:
Yes, these masks should have values between zero and one. They are defined in function of the magnitudes, so yes the log might be making things weird. 
I agree with Dan, different people will want different masks. Maybe having a method to obtain the masks from the mixture, the sources and the noise would make this choice easier when subclassing?

Reply:
I don't mind, but as Dan suggested in another comment, let's sort this out when building an actual source separation recipe and extend as we go. I changed the mask computation so that we first `exp` the features. You can check out how the masks look here (they're 0 - 1 range now). 

https://github.com/pzelasko/lhotse/blob/f2200fa7fe60fa134d0f256dd947aea3819f50de/examples/librimix/librimix.ipynb